### PR TITLE
MESH-1314/ [Staking] Permissioned entity

### PR DESCRIPTION
**Pull request by satyamakgec**

### CODE REVIEW

File: pallets/staking/src/lib.rs [hunk 1]
```diff
@@ -3581,12 +3593,16 @@ impl<T: Trait> Module<T> {
...
 
-    /// Is the stash account one of the permissioned validators?
-    pub fn is_validator_or_nominator_compliant(stash: &T::AccountId) -> bool {
-        if let Some(validator_identity) = <identity::Module<T>>::get_identity(&stash) {
-            return <identity::Module<T>>::has_valid_cdd(validator_identity);
-        }
-        false
+    /// Is nominator's stash account is compliant or not.
+    pub fn is_nominator_compliant(stash: &T::AccountId) -> bool {
+        <Identity<T>>::get_identity(&stash).map_or(false, |id| <Identity<T>>::has_valid_cdd(id))
```
Comment (Centril) [inline pallets/staking/src/lib.rs:3598]:
```suggestion
        <Identity<T>>::get_identity(&stash).map_or(false, <Identity<T>>::has_valid_cdd)
```

File: pallets/staking/src/lib.rs [hunk 2]
```diff
@@ -2081,40 +2095,39 @@ decl_module! {
...
-            let caller = ensure_signed(origin)?;
-            let caller_id = Context::current_identity_or::<T::Identity>(&caller).ok();
-            ensure!(Self::permissioned_validators(&validator), Error::<T>::NotExists);
-            // Change validator status to be Non-Permissioned
-            <PermissionedValidators<T>>::insert(&validator, false);
+        pub fn remove_permissioned_validator_entity(origin, entity: IdentityId) {
+            T::RequiredRemoveOrigin::ensure_origin(origin)?;
+            ensure!(Self::permissioned_entities(&entity), Error::<T>::NotExists);
+            // Change entity status to be Non-Permissioned
+            <PermissionedEntities>::insert(&entity, false);
```
Comment (Centril) [inline pallets/staking/src/lib.rs:2128]:
```suggestion
            PermissionedEntities::insert(&entity, false);
```

File: pallets/staking/src/lib.rs [hunk 3]
```diff
@@ -2081,40 +2095,39 @@ decl_module! {
...
-            ensure!(!Self::permissioned_validators(&validator), Error::<T>::AlreadyExists);
-            // Change validator status to be Permissioned
-            <PermissionedValidators<T>>::insert(&validator, true);
-            let validator_id = <identity::Module<T>>::get_identity(&validator);
-            Self::deposit_event(RawEvent::PermissionedValidatorAdded(validator_id, validator));
+            ensure!(!Self::permissioned_entities(&entity), Error::<T>::AlreadyExists);
+            // Validate the cdd status of the entity.
+            ensure!(<Identity<T>>::has_valid_cdd(entity), Error::<T>::InvalidEntityIdentity);
+            // Change entity status to be Permissioned
+            <PermissionedEntities>::insert(&entity, true);
```
Comment (Centril) [inline pallets/staking/src/lib.rs:2112]:
```suggestion
            PermissionedEntities::insert(&entity, true);
```

File: pallets/staking/src/lib.rs [hunk 4]
```diff
@@ -1497,7 +1505,13 @@ decl_error! {
         /// Updates with same value.
         NoChange,
         /// Updates with same value.
-        InvalidCommission
+        InvalidCommission,
+        /// Given potential entity has invalid identity.
+        InvalidEntityIdentity,
+        /// Stash is not a part of any allowed entities.
+        StashNotAllowed,
+        /// Stash doesn't have a identityId.
```
Comment (Centril) [inline pallets/staking/src/lib.rs:1513]:
```suggestion
        /// Stash doesn't have a DID.
```

File: pallets/staking/src/lib.rs [hunk 5]
```diff
@@ -1497,7 +1505,13 @@ decl_error! {
         /// Updates with same value.
         NoChange,
         /// Updates with same value.
-        InvalidCommission
+        InvalidCommission,
+        /// Given potential entity has invalid identity.
```
Comment (Centril) [inline pallets/staking/src/lib.rs:1509]: Not sure what this means; can you rephrase?

File: pallets/staking/src/lib.rs [hunk 6]
```diff
@@ -1315,9 +1317,9 @@ decl_storage! {
         /// forcing into account.
         pub IsCurrentSessionFinal get(fn is_current_session_final): bool = false;
 
-        /// The map from (wannabe) validators to the status of compliance.
-        pub PermissionedValidators get(fn permissioned_validators):
-            map hasher(twox_64_concat) T::AccountId => bool;
+        /// Allowed entities who shows interests to run operators/validator node.
```
Comment (Centril) [inline pallets/staking/src/lib.rs:1320]:
```suggestion
        /// Allowed entities who shows interests in running operators/validator node.
```

File: pallets/runtime/tests/src/staking/mod.rs [hunk 7]
```diff
@@ -5410,3 +5431,45 @@ fn voting_for_pip_overlays_with_staking() {
...
+        .validator_count(5)
+        .minimum_validator_count(5)
+        .build()
+        .execute_with(|| {
+            start_era(1);
+
+            // add new validator
+            bond_validator(50, 51, 500000);
+
+            // Add other stash and cntrl to the same did.
```
Comment (Centril) [inline pallets/runtime/tests/src/staking/mod.rs:5447]: cntrl?

File: pallets/runtime/tests/src/staking/mod.rs [hunk 8]
```diff
@@ -5231,16 +5231,25 @@ fn should_add_permissioned_validators() {
...
             frame_system::RawOrigin::Root.into(),
-            acc_20.clone()
+            Identity::get_identity(&acc_20).unwrap()
         ));
-        assert_eq!(Staking::permissioned_validators(acc_10), true);
-        assert_eq!(Staking::permissioned_validators(acc_20), true);
+        assert_eq!(
+            Staking::permissioned_entities(Identity::get_identity(&acc_10).unwrap()),
+            true
+        );
```
Comment (Centril) [inline pallets/runtime/tests/src/staking/mod.rs:5248]: Let's make a function out of this, taking the AccountId + the true/false RHS of assert_eq and refactor below.

File: pallets/runtime/tests/src/staking/mod.rs [hunk 9]
```diff
@@ -397,9 +397,9 @@ fn staking_should_work() {
                 1500,
                 RewardDestination::Controller
             ));
-            assert_ok!(Staking::add_permissioned_validator(
+            assert_ok!(Staking::add_permissioned_validator_entity(
                 frame_system::RawOrigin::Root.into(),
-                3
+                Identity::get_identity(&3).unwrap()
             ));
```
Comment (Centril) [inline pallets/runtime/tests/src/staking/mod.rs:403]: Could refactor logic from here with a function as well.

File: pallets/staking/src/lib.rs (hunk 6 above)
Comment (maxsam4) [reply pallets/staking/src/lib.rs:1320]: I'd rephrase this to something like "Entities that are allowed to run operator/validator nodes".

File: pallets/runtime/tests/src/staking/mod.rs (hunk 7 above)
Comment (maxsam4) [reply pallets/runtime/tests/src/staking/mod.rs:5447]: I wonder if Satyam meant centril or control.

Comment (Centril) [reply pallets/runtime/tests/src/staking/mod.rs:5447]: Probably "controller" 😜

Comment (maxsam4) [reply pallets/runtime/tests/src/staking/mod.rs:5447]: neh, that's too different. I am leaning towards centril. :stuck_out_tongue_winking_eye:

Comment (satyamakgec) [reply pallets/runtime/tests/src/staking/mod.rs:5447]: It refers to the controller but I think centril is sounded better 😜

File: pallets/staking/src/lib.rs (hunk 5 above)
Comment (satyamakgec) [reply pallets/staking/src/lib.rs:1509]: I will fix this in Commission PR #670

File: pallets/runtime/tests/src/staking/mod.rs [hunk 10]
```diff
@@ -37,6 +37,33 @@ macro_rules! assert_session_era {
...
+            Staking::permissioned_entities(Identity::get_identity($acc_id).unwrap()),
+            false
+        );
+    };
+}
+
+macro_rules! assert_add_permissioned_validator {
+    ($acc_id:expr) => {
+        assert_ok!(Staking::add_permissioned_validator_entity(
+            frame_system::RawOrigin::Root.into(),
```
Comment (Centril) [inline pallets/runtime/tests/src/staking/mod.rs:61]:
```suggestion
            root(),
```
(`use super::committee_test::root`)

File: pallets/runtime/tests/src/staking/mod.rs [hunk 11]
```diff
@@ -37,6 +37,33 @@ macro_rules! assert_session_era {
...
+            Identity::get_identity($acc_id).unwrap()
+        ));
+    };
+}
```
Comment (Centril) [inline pallets/runtime/tests/src/staking/mod.rs:65]: These don't need to be macros; they can be functions using `#[track_caller]`.

File: pallets/staking/src/lib.rs [hunk 12]
```diff
@@ -1543,6 +1559,23 @@ decl_module! {
 
         fn deposit_event() = default;
 
+        fn on_runtime_upgrade() -> Weight {
+            use frame_support::storage::migration::{ StorageIterator, put_storage_value };
+
+            if StorageVersion::get() == Releases::V4_0_0 {
+                StorageIterator::<()>::new(b"Staking", b"PermissionedValidators")
```
Comment (Centril) [inline pallets/staking/src/lib.rs:1566]: Let's add a version of `migrate_double_map_keys` for single maps and use it here.

File: pallets/runtime/tests/src/staking/mod.rs (hunk 10 above)
Comment (satyamakgec) [reply pallets/runtime/tests/src/staking/mod.rs:61]: It doesn't give anything extra?? Although It is better to move `root()` from `committee_test` to storage.rs.

File: pallets/runtime/tests/src/staking/mod.rs (hunk 11 above)
Comment (satyamakgec) [reply pallets/runtime/tests/src/staking/mod.rs:65]: Is there any performance or security issue?

File: pallets/runtime/tests/src/staking/mod.rs (hunk 10 above)
Comment (Centril) [reply pallets/runtime/tests/src/staking/mod.rs:61]: No, its just shorter; Feel free to move it though.

File: pallets/runtime/tests/src/staking/mod.rs (hunk 11 above)
Comment (Centril) [reply pallets/runtime/tests/src/staking/mod.rs:65]: Well compile time perf; its less code to typecheck and stuff, but also macros give you better type checking errors, etc.

File: pallets/staking/src/lib.rs (hunk 12 above)
Comment (vkomenda) [reply pallets/staking/src/lib.rs]: You can use `StorageKeyIterator` that decodes the keys itself and doesn't require reversing or decoding inline.

File: pallets/staking/src/lib.rs [hunk 13]
```diff
@@ -1543,6 +1559,23 @@ decl_module! {
...
+                    .drain()
+                    .filter_map(|(key, value)| {
+                        T::AccountId::decode(&mut Twox64Concat::reverse(&key)).map_or(None, |id| {
+                            Some((<Identity<T>>::get_identity(&id).unwrap_or_default(), value))
+                        })
+                    })
+                    .for_each(|(key, value)| put_storage_value(b"Staking", b"PermissionedEntities", &key.as_bytes(), value));
```
Comment (vkomenda) [inline pallets/staking/src/lib.rs]: You could either type `key` and `value` explicitly or rename those to something that tells about their types, possibly `did` and `account_id`?

File: pallets/staking/src/lib.rs [hunk 14]
```diff
@@ -2081,40 +2114,39 @@ decl_module! {
             ValidatorCount::mutate(|n| *n += factor * *n);
         }
 
-        /// Governance committee on 2/3 rds majority can introduce a new potential validator
-        /// to the pool of validators. Staking module uses `PermissionedValidators` to ensure
-        /// validators have completed KYB compliance and considers them for validation.
+        /// Governance committee on 2/3 rds majority can introduce a new potential entity
+        /// to the pool of complianced entities who can run validators. Staking module uses `PermissionedEntities`
```
Comment (vkomenda) [inline pallets/staking/src/lib.rs:2113]: "complianced entities" -> "permissioned entities"

File: pallets/staking/src/lib.rs [hunk 15]
```diff
@@ -2081,40 +2114,39 @@ decl_module! {
...
+        /// to ensure validators have completed KYB compliance and considers them for validation.
```
Comment (vkomenda) [inline pallets/staking/src/lib.rs:2114]: "KYB" -> "KYC"?

File: pallets/staking/src/lib.rs [hunk 16]
```diff
@@ -2081,40 +2114,39 @@ decl_module! {
...
+            // Validate the cdd status of the entity.
+            ensure!(<Identity<T>>::has_valid_cdd(entity), Error::<T>::InvalidEntityIdentity);
+            // Change entity status to be Permissioned
+            PermissionedEntities::insert(&entity, true);
+            Self::deposit_event(RawEvent::PermissionedEntityAdded(GC_DID, entity));
         }
 
-        /// Remove a validator from the pool of validators. Effects are known in the next session.
-        /// Staking module checks `PermissionedValidators` to ensure validators have
+        /// Remove an entity from the pool of (wannable) entity validators. Effects are known in the next session.
```
Comment (vkomenda) [inline pallets/staking/src/lib.rs:2130]: "(wannable)" -> "(wannabe)" or "aspiring"?

File: primitives/src/migrate.rs [hunk 17]
```diff
@@ -87,6 +87,37 @@ pub fn migrate_map_rename<T: Migrate, C: FnMut(&[u8]) -> T::Context>(
...
+    mut map: F,
+) where
+    F: FnMut(KO, VO) -> (KN, VN),
+    VO: Decode + Encode,
+    VN: Encode + Decode,
+    H: ReversibleStorageHasher,
+    KO: Decode,
+    KN: Decode + Encode,
+{
+    let old_map = StorageIterator::<VO>::new(module, item)
```
Comment (vkomenda) [inline primitives/src/migrate.rs:106]: `StorageKeyIterator` would be better in that it does reversing and decoding internally.

File: pallets/staking/src/lib.rs [hunk 18]
```diff
@@ -1356,10 +1361,13 @@ decl_storage! {
...
                         ).expect("Unable to add to Validator list");
-                        <Module<T>>::add_permissioned_validator(
-                            frame_system::RawOrigin::Root.into(),
-                            stash.clone()
-                        )
+                        if !<Module<T>>::permissioned_entities(entity_id) {
+                            let _ = <Module<T>>::add_permissioned_validator_entity(
+                                frame_system::RawOrigin::Root.into(),
+                                entity_id
+                            );
```
Comment (maxsam4) [inline pallets/staking/src/lib.rs:1368]: [](url)
```suggestion
                            <Module<T>>::add_permissioned_validator_entity(
                                frame_system::RawOrigin::Root.into(),
                                entity_id
                            ).expect("Unable to add to Permissioned entity list");
```

File: pallets/staking/src/lib.rs (hunk 15 above)
Comment (satyamakgec) [reply pallets/staking/src/lib.rs:2114]: It is fine KYB stands for KnowYourBusiness

**Review by vkomenda — APPROVED** [review]

**Review by maxsam4 — APPROVED** [review]

File: pallets/staking/src/lib.rs [hunk 19]
```diff
@@ -98,7 +98,7 @@
 //! An account can become a validator candidate via the
 //! [`validate`](./enum.Call.html#variant.validate) call
 //! But only those validators are in effect whose compliance status is active via
-//! [`add_permissioned_validator`](./enum.Call.html#variant.validate) call & there _stash_ accounts has valid CDD claim.
+//! [`add_permissioned_validator_entity`](./enum.Call.html#variant.validate) call & there _stash_ accounts has valid CDD claim.
```
Comment (adamdossa) [inline pallets/staking/src/lib.rs:101]: nit - can we keep the old name - entity is really the PUIS ID granularity.

File: pallets/staking/src/lib.rs [hunk 20]
```diff
@@ -1315,9 +1319,9 @@ decl_storage! {
         /// forcing into account.
         pub IsCurrentSessionFinal get(fn is_current_session_final): bool = false;
 
-        /// The map from (wannabe) validators to the status of compliance.
-        pub PermissionedValidators get(fn permissioned_validators):
-            map hasher(twox_64_concat) T::AccountId => bool;
+        /// Entities that are allowed to run operator/validator nodes.
+        pub PermissionedEntities get(fn permissioned_entities):
```
Comment (adamdossa) [inline pallets/staking/src/lib.rs:1323]: I think we should stick w/ validator for these.

**Review by adamdossa — APPROVED** [review]

File: pallets/staking/src/lib.rs (hunk 20 above)
Comment (satyamakgec) [reply pallets/staking/src/lib.rs:1323]: In a code sense, `IdentityId` doesn't represent the validator It is only the representation of the company/project/entity. I don't think it is reasonable to keep the name validator for the storage. If we don't like entity as a suffix then we can use another.

File: pallets/staking/src/lib.rs (hunk 19 above)
Comment (satyamakgec) [reply pallets/staking/src/lib.rs:101]: We are not really permission for the validator directly. But if you think it is better to keep the same name then I can do it.

File: pallets/staking/src/lib.rs (hunk 20 above)
Comment (adamdossa) [reply pallets/staking/src/lib.rs:1323]: Maybe `PermissionedIdentity`? On our terminology page we use:
Identity - relates to the on-chain identity - an entity can have multiple identities
Entity - represents a single "real-life" entity, that can have multiple identities.

Comment (satyamakgec) [reply pallets/staking/src/lib.rs:1323]: Yeah that works
